### PR TITLE
Initial commit for embedding JerryScript to specific target boards.

### PR DESCRIPTION
**Pull request by seanshpark**
This change includes mbed/frdm-k64f and esp8266-01.

JerryScript-DCO-1.0-Signed-off-by: SaeHie Park saehie.park@samsung.com

### CODE REVIEW

**Comment by egavrin** [comment]
Please split your commit into 3 commits:
- Infrastructural changes
- +ESP
- +K65F

**Comment by egavrin** [comment]
Please rename `embedding` into `targets`

File: embedding/esp8266/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+user/esp8266_js.h
```
Comment (egavrin) [inline embedding/esp8266/.gitignore]: Do we need this file here?

File: embedding/mbedk64f/.gitignore [hunk 2]
```diff
@@ -0,0 +1,5 @@
+libjerry/
```
Comment (egavrin) [inline embedding/mbedk64f/.gitignore]: Do we need this file here?

**Comment by egavrin** [comment]
Please share guide on how to build engine/firmware for K64F.

**Comment by egavrin** [comment]
Do you plan to more STM32* build to this infra?

File: embedding/esp8266/readme.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+Files in this folder (embedding/esp8266) are copied from
+examples/project_template of esp_iot_rtos_sdk and modified for JerryScript.
+You can view online from [this](https://github.com/espressif/esp_iot_rtos_sdk/tree/master/examples/project_template) page.
+
+
+### How to build JerryScript for ESP8266
+
+#### 1. SDK
+
+Follow [this](https://github.com/seanshpark/esp8266-docs/wiki/Preparation-and-setting-build-environment)
```
Comment (egavrin) [inline embedding/esp8266/readme.md]: Great guide! What do this about moving it into root source directory of the project? i.e `./docs/ESP-PREREQUISITES.md`

File: jerry-core/ecma/base/ecma-globals.h [hunk 4]
```diff
@@ -765,11 +765,11 @@ typedef struct
...
 typedef struct
 {
-  /** Compressed pointer to next chunk */
-  mem_cpointer_t next_chunk_cp;
-
   /** Characters */
   lit_utf8_byte_t data[ sizeof (uint64_t) - sizeof (mem_cpointer_t) ];
+
+  /** Compressed pointer to next chunk */
+  mem_cpointer_t next_chunk_cp;
```
Comment (egavrin) [inline jerry-core/ecma/base/ecma-globals.h]: Please move this fix into a separate commit.

File: jerry-core/jrt/jrt.h [hunk 5]
```diff
@@ -120,7 +120,11 @@ extern void __noreturn jerry_unimplemented (const char *, const char *, const ch
 #define JERRY_DDDLOG(...) JERRY_DLOG (__VA_ARGS__)
 #endif /* !JERRY_ENABLE_LOG */
 
+#if defined (__TARGET_ESP8266)
+#define JERRY_ERROR_MSG(...) printf (__VA_ARGS__)
```
Comment (egavrin) [inline jerry-core/jrt/jrt.h]: I think we should keep core sources as much target independent as possible, please implement `fprintf` in ESP8266 sources.

File: jerry-core/lit/lit-strings.cpp [hunk 6]
```diff
@@ -1005,11 +1005,19 @@ lit_put_ecma_char (ecma_char_t ecma_char) /**< code unit */
 {
   if (ecma_char <= LIT_UTF8_1_BYTE_CODE_POINT_MAX)
   {
-    putchar (ecma_char);
+#if defined (__TARGET_ESP8266)
```
Comment (egavrin) [inline jerry-core/lit/lit-strings.cpp]: Likewise.

File: jerry-core/lit/lit-strings.cpp [hunk 7]
```diff
@@ -1005,11 +1005,19 @@ lit_put_ecma_char (ecma_char_t ecma_char) /**< code unit */
...
+    printf ("%c", ecma_char);
+#else
+    putchar ((char) ecma_char);
+#endif
   }
   else
   {
     FIXME ("Support unicode characters printing.");
+#if defined (__TARGET_ESP8266)
+    printf ("_");
```
Comment (egavrin) [inline jerry-core/lit/lit-strings.cpp]: Likewise.

File: jerry-core/lit/lit-strings.cpp (hunk 6 above)
Comment (egavrin) [inline jerry-core/lit/lit-strings.cpp]: Actually, this printf shouldn't be here - it looks like issue.

File: jerry-core/lit/lit-strings.cpp [hunk 8]
```diff
@@ -1005,11 +1005,19 @@ lit_put_ecma_char (ecma_char_t ecma_char) /**< code unit */
 {
   if (ecma_char <= LIT_UTF8_1_BYTE_CODE_POINT_MAX)
   {
-    putchar (ecma_char);
+#if defined (__TARGET_ESP8266)
+    printf ("%c", ecma_char);
+#else
+    putchar ((char) ecma_char);
```
Comment (egavrin) [inline jerry-core/lit/lit-strings.cpp]: Likewise.

File: embedding/esp8266/.gitignore (hunk 1 above)
Comment (seanshpark) [inline embedding/esp8266/.gitignore]: this file is generated from js soure files to embed into memory.

File: embedding/mbedk64f/.gitignore (hunk 2 above)
Comment (seanshpark) [inline embedding/mbedk64f/.gitignore]: this folder is where jerry libraries are copied, so I think it won't be necessary to include to repo

**Comment by seanshpark** [comment]
> Please share guide on how to build engine/firmware for K64F.
`readme.md` file in `mbedk64f` has a link, http://yottadocs.mbed.com/#installing. maybe thinking of some short version?

**Comment by seanshpark** [comment]
> Do you plan to move STM32* build to this infra?

You mean current bare bone version? well, no. When this target embedding structure be stable and lands,  I would like someone interested can do.

File: embedding/esp8266/readme.md (hunk 3 above)
Comment (seanshpark) [inline embedding/esp8266/readme.md]: Thank you :). Yes, I need this kind of comments. I'll put a copy.

File: jerry-core/ecma/base/ecma-globals.h (hunk 4 above)
Comment (seanshpark) [inline jerry-core/ecma/base/ecma-globals.h]: OK

File: jerry-core/jrt/jrt.h (hunk 5 above)
Comment (seanshpark) [inline jerry-core/jrt/jrt.h]: Yes, I totally agree. `fprintf` exist in ESP8266 SDK `libcirom.a`, but I have `undefined reference to `_impure_ptr'` link error. I need some help one this.

File: jerry-core/lit/lit-strings.cpp (hunk 8 above)
Comment (seanshpark) [inline jerry-core/lit/lit-strings.cpp]: OK, an issue we can talk about :)

**Comment by seanshpark** [comment]
I'll update this change when #752 is settled down.

**Comment by seanshpark** [comment]
> Please split your commit into 3 commits

Sure, it's important to make one thing at a time.

File: embedding/esp8266/.gitignore (hunk 1 above)
Comment (egavrin) [inline embedding/esp8266/.gitignore]: I mean `.gitignore`?

Comment (seanshpark) [inline embedding/esp8266/.gitignore]: yes, to ignore build time generated files and it be dependent on targets. can I ask if you have some other things in mind?

File: CMakeLists.txt [hunk 9]
```diff
@@ -419,11 +425,11 @@ endif()
     target_include_directories(${TARGET_NAME} SYSTEM PRIVATE ${INCLUDE_EXTERNAL_LIBS_INTERFACE})
     if(("${PLATFORM}" STREQUAL "DARWIN") AND (NOT (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX)))
      target_link_libraries(${TARGET_NAME} ${CORE_TARGET_NAME} ${LIBC_TARGET_NAME}
-		                   ${FDLIBM_TARGET_NAME} ${PREFIX_IMPORTED_LIB}libclang_rt.osx)
-	else()
+                           ${FDLIBM_TARGET_NAME} ${PREFIX_IMPORTED_LIB}libclang_rt.osx)
```
Comment (egavrin) [inline CMakeLists.txt]: Accident reformatting?

File: CMakeLists.txt [hunk 10]
```diff
@@ -419,11 +425,11 @@ endif()
...
+    else()
      target_link_libraries(${TARGET_NAME} ${CORE_TARGET_NAME} ${LIBC_TARGET_NAME}
-		                   ${FDLIBM_TARGET_NAME} ${PREFIX_IMPORTED_LIB}libgcc ${PREFIX_IMPORTED_LIB}libgcc_eh)
-	endif()
+                           ${FDLIBM_TARGET_NAME} ${PREFIX_IMPORTED_LIB}libgcc ${PREFIX_IMPORTED_LIB}libgcc_eh)
```
Comment (egavrin) [inline CMakeLists.txt]: Likewise

File: CMakeLists.txt [hunk 11]
```diff
@@ -501,7 +519,7 @@ endif()
     if(("${PLATFORM}" STREQUAL "DARWIN") AND (NOT (CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX)))
       target_link_libraries(${TARGET_NAME} ${CORE_TARGET_NAME} ${LIBC_TARGET_NAME} ${FDLIBM_TARGET_NAME}
                             ${PREFIX_IMPORTED_LIB}libclang_rt.osx)
-	else()
+    else()
```
Comment (egavrin) [inline CMakeLists.txt]: Likewise

File: embedding/esp8266/.gitignore (hunk 1 above)
Comment (egavrin) [inline embedding/esp8266/.gitignore]: I mean `.gitignore` file itself. We can update the root `.gitignore`?

File: embedding/esp8266/Makefile.esp8266 [hunk 12]
```diff
@@ -0,0 +1,103 @@
+# Copyright 2014-2015 Samsung Electronics Co., Ltd.
```
Comment (egavrin) [inline embedding/esp8266/Makefile.esp8266]: Years of copyright are incorrect, should be `2015`.

File: embedding/esp8266/js/main.js [hunk 13]
```diff
@@ -0,0 +1,4 @@
+function sysloop(ticknow) {
```
Comment (egavrin) [inline embedding/esp8266/js/main.js]: I'm worrying about the fact that we have different hello-world applications for different platforms. We can improve it later, just to think.

File: embedding/mbedk64f/js2c.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
```
Comment (egavrin) [inline embedding/mbedk64f/js2c.py]: I accidentaly realised that `js2c.py` is doing same thing as `./tools/generator.sh`. Additionally, `./embedding/mbedk64f/js2c.py` and `./embedding/esp8266/js2c.py` looks nearly the same.

File: CMakeLists.txt (hunk 9 above)
Comment (seanshpark) [inline CMakeLists.txt]: It's removal of TAB. This part is already landed by https://github.com/Samsung/jerryscript/commit/5ffc9abb592ba6072a100b2c73de09c6623f4032

File: embedding/esp8266/js/main.js (hunk 13 above)
Comment (seanshpark) [inline embedding/esp8266/js/main.js]: Yes agree. It's same LED blink example and I want to make them work with one js source code. Maybe move them to some common folder and edit the Makefile for both. I'll work on this after this PR lands.

File: embedding/mbedk64f/js2c.py (hunk 14 above)
Comment (seanshpark) [inline embedding/mbedk64f/js2c.py]: Yes with some file path difference. This seems to be done with making common js example code.

File: embedding/esp8266/.gitignore (hunk 1 above)
Comment (seanshpark) [inline embedding/esp8266/.gitignore]: yes, merge to root `.gitignore` is ok.

**Comment by seanshpark** [comment]
@egavrin , I think all of your comments are applied :)

**Comment by egavrin** [comment]
I have only minor question about merging `js2c.py` and `./tools/generator.sh`. 
Otherwise LGTM

**Comment by seanshpark** [comment]
Didn't see the last message :(
`about merging js2c.py and ./tools/generator.sh` 
I think it would be better to merge them when MCU moves to `targets` folder. But before that, is it ok to change MCU codes to `targets` style and build structure?

**Comment by seanshpark** [comment]
Waiting for one more `LGTM` :)

File: targets/mbedk64f/js2c.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+import re
+
+def extractName(path):
+    return os.path.splitext(os.path.basename(path))[0]
+
+def writeLine(fo, content, indent=0):
+    buf = '  ' * indent
+    buf += content
+    buf += '\n'
+    fo.write(buf)
```
Comment (galpeter) [inline targets/mbedk64f/js2c.py]: Why don't we simply write out everything directly? Do we really need to build up a string with `+` ?

File: targets/mbedk64f/js2c.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+OUT_PATH = './source/'
+SRC_PATH = './js/'
+
+# argument processing
+buildtype = 'release'
+if len(sys.argv) >= 2:
+    buildtype = sys.argv[1]
+
+fout = open(OUT_PATH + 'jerry_targetjs.h', 'w')
```
Comment (galpeter) [inline targets/mbedk64f/js2c.py]: Should the `fout` file object be closed after every write is done?

File: targets/mbedk64f/js2c.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+fout.write(LICENSE);
+fout.write(HEADER);
+
+def export_one_file(path, name):
```
Comment (galpeter) [inline targets/mbedk64f/js2c.py]: In this file we are mixing name conventions for methods, I would prefer to use one.

File: targets/mbedk64f/js2c.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    fout.write('const static char ' + name + '_n[] = "' + name + '";\n')
+    fout.write('const static char ' + name + '_s[] =\n{\n')
+
+    code = open(path, 'r').read() + '\0'
```
Comment (galpeter) [inline targets/mbedk64f/js2c.py]: same as `fout` the file is not closed at all. If we are certain that the python version is at least 2.6 in all cases, then this could be written like this:
```
with open(path, 'r') as code_file:
    code = code_file.read() + '\0'
```

BTW: do we really need the '\0'?

File: targets/mbedk64f/source/jerry_extapi.cpp [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+  {
+    printf(">> print(%d) :", (int)args_cnt);
+    for (cc=0; cc<args_cnt; cc++)
+    {
+      if (args_p[cc].type == JERRY_API_DATA_TYPE_STRING && args_p[cc].v_string)
+      {
+        static char buffer[128];
+        int length, maxlength;
+        length = -jerry_api_string_to_char_buffer (args_p[0].v_string, NULL, 0);
+        maxlength  = MAX(length, 126);
```
Comment (galpeter) [inline targets/mbedk64f/source/jerry_extapi.cpp]: If the length is bigger than 128 then the max value will be that number and then we'll be copying more than the size of the buffer into the array, is that really correct? Shouldn't this be MIN instead?

**Comment by galpeter** [comment]
General question: do we plan to have the same Jerry code style requirements for the files under the `targets` dir?

File: targets/esp8266/gen_misc.sh [hunk 20]
```diff
@@ -0,0 +1,165 @@
+#!/bin/bash
```
Comment (galpeter) [inline targets/esp8266/gen_misc.sh]: Shouldn't this file have a copyright info?

File: targets/esp8266/gen_misc.sh [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+echo "spi mode: $spi_mode"
+echo ""
+
+echo "STEP 5: choose spi size and map"
+echo "    0= 512KB( 256KB+ 256KB)"
+echo "    2=1024KB( 512KB+ 512KB)"
+echo "    3=2048KB( 512KB+ 512KB)"
+echo "    4=4096KB( 512KB+ 512KB)"
+echo "    5=2048KB(1024KB+1024KB)"
+echo "    6=4096KB(1024KB+1024KB)"
```
Comment (galpeter) [inline targets/esp8266/gen_misc.sh]: Is the '1024+1024' correct here?

File: targets/esp8266/js2c.py [hunk 22]
```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
```
Comment (galpeter) [inline targets/esp8266/js2c.py]: It seems that this file is present in both targets. Do we really need it twice? Is it somehow target dependent?

File: targets/mbedk64f/js2c.py (hunk 15 above)
Comment (seanshpark) [inline targets/mbedk64f/js2c.py]: fix to one line :)

File: targets/mbedk64f/js2c.py (hunk 16 above)
Comment (seanshpark) [inline targets/mbedk64f/js2c.py]: `fout` close is missing, and source file close `code = open(path, 'r').read() + '\0'` is missing. needs fix.

File: targets/mbedk64f/js2c.py (hunk 17 above)
Comment (seanshpark) [inline targets/mbedk64f/js2c.py]: yes, agree

File: targets/mbedk64f/source/jerry_extapi.cpp (hunk 19 above)
Comment (seanshpark) [inline targets/mbedk64f/source/jerry_extapi.cpp]: yes, it should be MIN.

File: targets/esp8266/gen_misc.sh (hunk 20 above)
Comment (seanshpark) [inline targets/esp8266/gen_misc.sh]: it's from the esp8266 code and there is no copyright info and no changes.

File: targets/esp8266/js2c.py (hunk 22 above)
Comment (seanshpark) [inline targets/esp8266/js2c.py]: js2c.py is duplicate and somehow I will merge it to one after this PR :)

**Comment by seanshpark** [comment]
> General question: do we plan to have the same Jerry code style requirements for the files under the targets dir?

Files from the target provider may or should co-exist and it can have its own code style. I think it would be comfortable to leave it as by targets style.

File: targets/mbedk64f/js2c.py (hunk 18 above)
Comment (wateret) [inline targets/mbedk64f/js2c.py]: @galpeter I think we need `'\0'` because we are building a character array like `'a', 'b', 'c', ...` not `"abc..."`

**Comment by seanshpark** [comment]
@galpeter , please review again when comfortable.

**Comment by galpeter** [comment]
@seanshpark LGTM
